### PR TITLE
Fix S3626 FP: Redundant jump for "throw expressions"

### LINKED ISSUE #3105: Fix S3626 FP: "throw" expression in a delegates
I have a Xunit test that looks like the following. In the arrange part of the test, I have a delegate `(httpContext) => throw exception` which gets flagged with "Remove this redundant jump." (S3626)`. In this test, I am expecting a certain behaviour when an exception is thrown.

I'm using `SonarAnalyzer.CSharp`, version `8.3.0.14607`.

```C#
[Fact]
public async Task When_InvokingMiddlewareWithException_Expect_LogLevelCritical()
{
    // Arrange
    Exception exception = new Exception("Unit Test");

    Mock<ILogger<RequestLoggingMiddleware>> mock = new Mock<ILogger<RequestLoggingMiddleware>>();
    mock.Setup(m => m.IsEnabled(It.IsAny<LogLevel>())).Returns(true);

    RequestLoggingMiddleware middleware = new RequestLoggingMiddleware(
        (httpContext) => throw exception,
        mock.Object);

    HttpContext context = new DefaultHttpContext();
    context.Response.Body = new MemoryStream();

    // Act
    Exception result = await Record.ExceptionAsync(() => middleware.Invoke(context));

    // Assert
    mock.Verify(
        m => m.Log(
            LogLevel.Information,
            It.IsAny<EventId>(),
            It.IsAny<It.IsAnyType>(),
            null,
            (Func<It.IsAnyType, Exception, string>)It.IsAny<object>()),
        Times.Once);

    mock.Verify(
        m => m.Log(
            LogLevel.Critical,
            It.IsAny<EventId>(),
            It.IsAny<It.IsAnyType>(),
            exception,
            (Func<It.IsAnyType, Exception, string>)It.IsAny<object>()),
        Times.Once);

    Assert.NotNull(result);
    Assert.Equal(exception, result);
}
```

A copy of class that the Xunit test tests is below.

```C#
namespace Scaffold.WebApi.Middleware
{
    using System;
    using System.Diagnostics;
    using System.Threading.Tasks;
    using Microsoft.AspNetCore.Http;
    using Microsoft.Extensions.Logging;

    public class RequestLoggingMiddleware
    {
        private const string RequestStartedMessageTemplate = "Inbound HTTP {HttpMethod} {Path} started";

        private const string RequestFinishedMessageTemplate = "Inbound HTTP {HttpMethod} {Path} finished in {ElapsedMilliseconds}ms - {StatusCode}";

        private static readonly Action<ILogger, string, PathString, Exception?> LogRequestStarted =
            LoggerMessage.Define<string, PathString>(LogLevel.Information, default, RequestStartedMessageTemplate);

        private static readonly Action<ILogger, string, PathString, long, int, Exception?> LogRequestFinishedInformation =
            LoggerMessage.Define<string, PathString, long, int>(LogLevel.Information, default, RequestFinishedMessageTemplate);

        private static readonly Action<ILogger, string, PathString, long, int, Exception?> LogRequestFinishedWarning =
            LoggerMessage.Define<string, PathString, long, int>(LogLevel.Warning, default, RequestFinishedMessageTemplate);

        private static readonly Action<ILogger, string, PathString, long, int, Exception?> LogRequestFinishedError =
            LoggerMessage.Define<string, PathString, long, int>(LogLevel.Error, default, RequestFinishedMessageTemplate);

        private static readonly Action<ILogger, string, PathString, long, string, Exception?> LogRequestFinishedCritical =
            LoggerMessage.Define<string, PathString, long, string>(LogLevel.Critical, default, RequestFinishedMessageTemplate);

        private readonly RequestDelegate next;

        private readonly ILogger logger;

        public RequestLoggingMiddleware(RequestDelegate next, ILogger<RequestLoggingMiddleware> logger)
        {
            this.next = next;
            this.logger = logger;
        }

        public async Task Invoke(HttpContext httpContext)
        {
            HttpRequest request = httpContext.Request;
            HttpResponse response = httpContext.Response;

            LogRequestStarted(this.logger, request.Method, request.Path, null);

            Stopwatch stopwatch = Stopwatch.StartNew();

            try
            {
                await this.next.Invoke(httpContext);
                stopwatch.Stop();

                if (response.StatusCode >= 200 && response.StatusCode <= 299)
                {
                    LogRequestFinishedInformation(this.logger, request.Method, request.Path, stopwatch.ElapsedMilliseconds, response.StatusCode, null);
                    return;
                }

                if (response.StatusCode >= 500)
                {
                    LogRequestFinishedError(this.logger, request.Method, request.Path, stopwatch.ElapsedMilliseconds, response.StatusCode, null);
                    return;
                }

                LogRequestFinishedWarning(this.logger, request.Method, request.Path, stopwatch.ElapsedMilliseconds, response.StatusCode, null);
            }
            catch (Exception exception)
            {
                stopwatch.Stop();

                LogRequestFinishedCritical(this.logger, request.Method, request.Path, stopwatch.ElapsedMilliseconds, "Unhandled Exception", exception);

                throw;
            }
        }
    }
}
```

The delegate `(httpContext) => throw exception` should not be flagged by (S3626) as it doesn't even jump. It's just a declared function.

### PR DESCRIPTION
**Pull request by lg2de**
Fixes #3105

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantJumpStatement.cs [hunk 1]
```diff
@@ -148,7 +148,7 @@ private static bool IsOnlyYieldBreak(JumpBlock jumpBlock, int yieldStatementCoun
 
         private static bool IsThrow(JumpBlock jumpBlock)
         {
-            return jumpBlock.JumpNode is ThrowStatementSyntax;
+            return jumpBlock.JumpNode is ThrowStatementSyntax || ThrowExpressionSyntaxWrapper.IsInstance(jumpBlock.JumpNode);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantJumpStatement.cs:151]: More lightweight solution:
```suggestion
            return jumpBlock.JumpNode.IsAnyKind(SyntaxKind.ThrowStatement, SyntaxKindEx.ThrowExpression);
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Hi @lg2de, 

Thank you for taking time to look at this issue. I left one improvement comment, but it should be good to go otherwise.

**Comment by lg2de** [comment]
I've committed your suggestion, @pavel-mikula-sonarsource . I did not know this option. Hope the tests will succeed too.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**